File: c#/tema_5/protobuf/ProtoUtils.cs
using System;
using Model;
using proto=Concertpro.Network.Protobuf;

namespace protobuf
{
    public class ProtoUtils
    {
        public static proto.Request CreateLoginRequest(Model.User user)
        {
            proto.User userDto = new proto.User {Id = user.Id, Username = user.Username, Password = user.Password, 
                LoggedIn = user.LoggedIn};
            proto.Request request = new proto.Request {Type = proto.Request.Types.Type.Login, User = userDto};

            return request;
        }

        public static proto.Response CreateOkResponse()
        {
            proto.Response response = new proto.Response {Type = proto.Response.Types.Type.Ok};
            return response;
        }
        
        public static proto.Response CreateErrorResponse(string text)
        {
            proto.Response response = new proto.Response {Type = proto.Response.Types.Type.Error, Error = text};
            return response;
        }

        public static proto.Request CreateAddOrderRequest(Order order)
        {
            proto.Order orderDto = new proto.Order
                { BuyerName = order.BuyerName, NumberOfSeats = order.NumberOfSeats };
            orderDto.Spectacle = new proto.Spectacle {Id = order.Spectacle.Id, 
                SpectaclePlace = order.Spectacle.SpectaclePlace, SpectacleDate = order.Spectacle.SpectacleDate.ToShortDateString(), 
                SeatsAvailable = order.Spectacle.SeatsAvailable, SeatsSold = order.Spectacle.SeatsSold, 
                ArtistName = order.Spectacle.ArtistName};
            proto.Request request = new proto.Request{Type = proto.Request.Types.Type.AddOrder, Order = orderDto};

            return request;
        }
        
        public static proto.Response CreateGetSpectaclesResponse(Model.Spectacle[] spectacles)
        {
            proto.Response response = new proto.Response {Type = proto.Response.Types.Type.GetSpectacles};
            foreach (Model.Spectacle spectacle in spectacles)
            {
                proto.Spectacle spectacleDto = new proto.Spectacle {Id = spectacle.Id, ArtistName = spectacle.ArtistName, 
                    SpectacleDate = spectacle.SpectacleDate.ToShortDateString(), SpectaclePlace = spectacle.SpectaclePlace, 
                    SeatsAvailable = spectacle.SeatsAvailable, SeatsSold = spectacle.SeatsSold};
                response.Spectacles.Add(spectacleDto);
            }

            return response;
        }

        public static proto.Response CreateGetFilteredSpectaclesResponse(Model.Spectacle[] spectacles)
        {
            proto.Response response = new proto.Response {Type = proto.Response.Types.Type.GetFilteredSpectacles};
            foreach (Model.Spectacle spectacle in spectacles)
            {
                proto.Spectacle spectacleDto = new proto.Spectacle {Id = spectacle.Id, SpectaclePlace = spectacle.SpectaclePlace,
                    ArtistName = spectacle.ArtistName, SpectacleDate = spectacle.SpectacleDate.ToShortDateString(), 
                    SeatsAvailable = spectacle.SeatsAvailable, SeatsSold = spectacle.SeatsSold};
                response.Spectacles.Add(spectacleDto);
            }

            return response;
        }

        public static string GetError(proto.Response response)
        {
            String errorMessage = response.Error;
            return errorMessage;
        }

        public static Model.User GetUser(proto.Response response)
        {
            Model.User user = new Model.User {Id = response.User.Id, Username = response.User.Username, 
                Password = response.User.Password, LoggedIn = response.User.LoggedIn};
            return user;
        }

        public static Model.User GetUser(proto.Request request)
        {
            Model.User user = new Model.User {Id = request.User.Id, Username = request.User.Username, 
                Password = request.User.Password, LoggedIn = request.User.LoggedIn};
            return user;
        }

        public static proto.Response AddOrderResponse()
        {
            proto.Response response = new proto.Response{ Type = proto.Response.Types.Type.AddedOrder };
            return response;
        }

        public static Model.Order GetOrder(proto.Request request)
        {
            Spectacle spectacle = new Spectacle
            {
                Id = request.Order.Spectacle.Id,
                ArtistName = request.Order.Spectacle.ArtistName,
                SpectacleDate = DateTime.Parse(request.Order.Spectacle.SpectacleDate),
                SpectaclePlace = request.Order.Spectacle.SpectaclePlace,
                SeatsAvailable = request.Order.Spectacle.SeatsAvailable,
                SeatsSold = request.Order.Spectacle.SeatsSold
            };
            Model.Order order = new Model.Order {BuyerName = request.Order.BuyerName, 
                Spectacle = spectacle, NumberOfSeats = request.Order.NumberOfSeats};
            return order;
        }

        public static DateTime GetDate(proto.Request request)
        {
            DateTime date = DateTime.Parse(request.Date);
            return date;
        }
    }
}

File: java/tema_6/ConcertProNetworking/src/main/java/concertpro/network/protobuf/ProtoUtils.java
package concertpro.network.protobuf;

import concertpro.model.Order;
import concertpro.model.Spectacle;
import concertpro.model.User;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ProtoUtils {
    public static ConcertProProtocol.Request createLoginRequest(User user){
        ConcertProProtocol.User userDto = ConcertProProtocol.User.newBuilder()
            .setUsername(user.getUsername())
            .setPassword(user.getPassword())
            .setLoggedIn(user.getLoggedIn())
            .build();

        ConcertProProtocol.Request request = ConcertProProtocol.Request.newBuilder()
            .setType(ConcertProProtocol.Request.Type.LOGIN)
            .setUser(userDto)
            .build();

        return request;
    }

    public static ConcertProProtocol.Request createGetSpectaclesRequest() {
        ConcertProProtocol.Request request = ConcertProProtocol.Request.newBuilder()
            .setType(ConcertProProtocol.Request.Type.GET_SPECTACLES)
            .build();
        return request;
    }

    public static ConcertProProtocol.Request createGetSpectaclesByDateRequest(LocalDate date) {
        ConcertProProtocol.Request request = ConcertProProtocol.Request.newBuilder()
            .setType(ConcertProProtocol.Request.Type.GET_FILTERED_SPECTACLES)
            .setDate(date.format(DateTimeFormatter.ofPattern("dd.MM.yyyy")))
            .build();

        return request;
    }

    public static ConcertProProtocol.Request createAddOrderRequest(Order order) {
        ConcertProProtocol.Spectacle spectacle = ConcertProProtocol.Spectacle.newBuilder()
            .setId(order.getSpectacle().getId())
            .setArtistName(order.getSpectacle().getArtistName())
            .setSpectacleDate(order.getSpectacle().getSpectacleDate().format(DateTimeFormatter.ofPattern("dd.MM.yyyy")))
            .setSpectaclePlace(order.getSpectacle().getSpectaclePlace())
            .setSeatsAvailable(order.getSpectacle().getSeatsAvailable())
            .setSeatsSold(order.getSpectacle().getSeatsSold())
            .build();

        ConcertProProtocol.Order orderDto = ConcertProProtocol.Order.newBuilder()
            .setBuyerName(order.getBuyerName())
            .setNumberOfSeats(order.getNumberOfSeats())
            .setSpectacle(spectacle)
            .build();

        ConcertProProtocol.Request request = ConcertProProtocol.Request.newBuilder()
            .setType(ConcertProProtocol.Request.Type.ADD_ORDER)
            .setOrder(orderDto)
            .build();

        return request;
    }

    public static Spectacle[] getSpectaclesResponse(ConcertProProtocol.Response response) {
        Spectacle[] spectacles = new Spectacle[response.getSpectaclesCount()];

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
        for(int i = 0; i < response.getSpectaclesCount(); i++) {
            ConcertProProtocol.Spectacle spectacleDto = response.getSpectacles(i);
            Spectacle spectacle = new Spectacle();
            spectacle.setId(spectacleDto.getId());
            spectacle.setArtistName(spectacleDto.getArtistName());
            spectacle.setSpectacleDate(LocalDate.parse(spectacleDto.getSpectacleDate(), formatter));
            spectacle.setSpectaclePlace(spectacleDto.getSpectaclePlace());
            spectacle.setSeatsAvailable(spectacleDto.getSeatsAvailable());
            spectacle.setSeatsSold(spectacleDto.getSeatsSold());
            spectacles[i] = spectacle;
        }

        return spectacles;
    }

    public static String getError(ConcertProProtocol.Response response){
        String errorMessage = response.getError();
        return errorMessage;
    }


}
